python_modules/libraries/dagster-databricks/dagster_databricks/databricks_pyspark_step_launcher.py

```diff
@@ -0,0 +1,229 @@
+        )
+        run_id = step_context.pipeline_run.run_id
+        log = step_context.log
+
+        step_key = step_run_ref.step_key
+        self._upload_artifacts(log, step_run_ref, run_id, step_key)
+
+        task = self._get_databricks_task(run_id, step_key)
+        databricks_run_id = self.databricks_runner.submit_run(self.run_config, task)
+
+        try:
+            self.databricks_runner.wait_for_run_to_complete(log, databricks_run_id)
+        finally:
+            if self.wait_for_logs:
+                self._log_logs_from_cluster(log, databricks_run_id)
+
+        for event in self.get_step_events(run_id, step_key):
+            log_step_event(step_context, event)
+            yield event
+
+    def get_step_events(self, run_id, step_key):
+        path = self._dbfs_path(run_id, step_key, PICKLED_EVENTS_FILE_NAME)
+        events_data = self.databricks_runner.client().read_file(path)
+        return deserialize_value(pickle.loads(events_data))
+
+    def _get_databricks_task(self, run_id, step_key):
+        """
+        Construct the 'task' parameter to  be submitted to the Databricks API.
+
+        This will create a 'spark_python_task' dict where `python_file` is a path on DBFS
+        pointing to the 'databricks_step_main.py' file, and `parameters` is an array with a single
+        element, a path on DBFS pointing to the picked `step_run_ref` data.
+
+        See https://docs.databricks.com/dev-tools/api/latest/jobs.html#jobssparkpythontask.
+        """
+        python_file = self._dbfs_path(run_id, step_key, self._main_file_name())
+        parameters = [
+            self._internal_dbfs_path(run_id, step_key, PICKLED_STEP_RUN_REF_FILE_NAME),
+            self._internal_dbfs_path(run_id, step_key, CODE_ZIP_NAME),
+            # TODO sd2k: remove the rest when dagster-azure/dagster-databricks are released
+            # and rely on libraries instead.
+            self._internal_dbfs_path(run_id, step_key, "dagster.zip"),
+            self._internal_dbfs_path(run_id, step_key, "dagster-aws.zip"),
+            # self._internal_dbfs_path(run_id, step_key, "dagster-azure.zip"),
+            self._internal_dbfs_path(run_id, step_key, "dagster-pyspark.zip"),
+            self._internal_dbfs_path(run_id, step_key, "dagster-spark.zip"),
+            self._internal_dbfs_path(run_id, step_key, "dagster-databricks.zip"),
+        ]
+        return {"spark_python_task": {"python_file": python_file, "parameters": parameters}}
+
+    def _upload_artifacts(self, log, step_run_ref, run_id, step_key):
+        """
+        Upload the step run ref and pyspark code to DBFS to run as a job.
+        """
+
+        log.info("Uploading main file to DBFS")
+        main_local_path = self._main_file_local_path()
+        with open(main_local_path, 'rb') as infile:
+            self.databricks_runner.client().put_file(
+                infile, self._dbfs_path(run_id, step_key, self._main_file_name())
+            )
+
+        if True:
```

Comment:
With EMR, we've basically punted this to the user, though I suppose that's more difficult with Databricks because the job is often launching the cluster itself instead of relying on a user to launch and configure it separately.

@natekupp - I'm curious whether you have thoughts on the right thing to do here.  I don't love this, but also don't have a better option in mind.

Reply:
> This is a really impressive and comprehensive piece of work. Thanks a ton for contributing.

Most of this was more or less copied from your work on the EMR subpackage so thanks for that! :slightly_smiling_face: 

> I left a few comments inside, mostly on stylistic stuff. I think the biggest open question I have is on how the storage piece fits in. Have you looked into whether it would make sense to expose the s3/dbfs referenced in the PR as an intermediate store?

I've pushed fixes for the stylistic stuff, thanks for the comments. Let me know if you'd like the changes squashing.

Storage is definitely a concern. I'm not convinced that it's worth adding another storage system for DBFS since they strongly recommend using either S3 or Azure anyway, plus they need to be accessed in different ways depending whether they're accessed from inside or outside the cluster. I've mentioned the use of an intermediate store in a separate comment above. Note that the launcher _does_ require and use a storage system (the `simple_pyspark` example uses S3), it just also requires additional credentials in Databricks so they can be mounted secretly.